### PR TITLE
odin: pin llvmPackages_18

### LINKED ISSUE #275919: Build failure: Odin
### Steps To Reproduce

1. build odin

### Build log

```console
[user@system:~]$ nix shell nixpkgs#odin
error: builder for '/nix/store/h8b0b2jxivnj6b4nx8lkkr8x14x7m9rs-odin-dev-2023-11.drv' failed with exit code 2;
       last 10 log lines:
       > patching script interpreter paths in build_odin.sh
       > build_odin.sh: interpreter directive changed from "#!/usr/bin/env sh" to "/nix/store/6nxav88iiz0g8m598xy643f8hhdz5kkx-bash-5.2-p21/bin/sh"
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: buildPhase
       > build flags: SHELL=/nix/store/6nxav88iiz0g8m598xy643f8hhdz5kkx-bash-5.2-p21/bin/bash release
       > ./build_odin.sh release
       > + clang++ src/main.cpp src/libtommath.cpp -Wno-switch -Wno-macro-redefined -Wno-unused-value '-DODIN_VERSION_RAW="dev-2023-12"' -std=c++14 -I/nix/store/b4w5ndycmgbhzjsa7nnbdimd9ybf1crb-llvm-13.0.1-dev/include -std=c++14 -stdlib=libc++ -fno-exceptions -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -L/nix/store/xq81zs44gk5xf6nqbxxqjdnv0b8fqs8a-llvm-13.0.1-lib/lib -Wl,-search_paths_first -Wl,-headerpad_max_install_names -O3 -pthread -lm -lstdc++ -liconv -ldl -lSystem -lLLVM -o odin
       > + set +x
       > ./build_odin.sh: line 119: 60617 Segmentation fault: 11  ./odin run examples/demo/demo.odin -file
       > make: *** [Makefile:13: release] Error 139
       For full logs, run 'nix-store -l /nix/store/h8b0b2jxivnj6b4nx8lkkr8x14x7m9rs-odin-dev-2023-11.drv'.
```

### Additional context

If you run the resulting binary through `lldb` you'll see this strange backtrace:
```console
lldb odin
(lldb) target create "odin"
Current executable set to '/Users/temp/Documents/personal/Odin/odin' (arm64).
(lldb) r
Process 56254 launched: '/Users/temp/Documents/personal/Odin/odin' (arm64)
Process 56254 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x0000000000000000
error: memory read failed for 0x0
Target 0: (odin) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x0000000000000000
    frame #1: 0x0000000100f1dbb8 libc++.1.0.dylib`std::__1::__stdinbuf<char>::imbue(std::__1::locale const&) + 52
    frame #2: 0x0000000100f1d41c libc++.1.0.dylib`std::__1::DoIOSInit::DoIOSInit() + 148
    frame #3: 0x0000000100f1e9d4 libc++.1.0.dylib`_GLOBAL__I_000100 + 72
    frame #4: 0x00000001802f1a24 dyld`invocation function for block in dyld4::Loader::findAndRunAllInitializers(dyld4::RuntimeState&) const::$_0::operator()() const + 168
    frame #5: 0x00000001803370f4 dyld`invocation function for block in dyld3::MachOAnalyzer::forEachInitializer(Diagnostics&, dyld3::MachOAnalyzer::VMAddrConverter const&, void (unsigned int) block_pointer, void const*) const + 172
    frame #6: 0x000000018032a668 dyld`invocation function for block in dyld3::MachOFile::forEachSection(void (dyld3::MachOFile::SectionInfo const&, bool, bool&) block_pointer) const + 496
    frame #7: 0x00000001802d12fc dyld`dyld3::MachOFile::forEachLoadCommand(Diagnostics&, void (load_command const*, bool&) block_pointer) const + 300
    frame #8: 0x00000001803296a0 dyld`dyld3::MachOFile::forEachSection(void (dyld3::MachOFile::SectionInfo const&, bool, bool&) block_pointer) const + 192
    frame #9: 0x000000018032c188 dyld`dyld3::MachOFile::forEachInitializerPointerSection(Diagnostics&, void (unsigned int, unsigned int, bool&) block_pointer) const + 160
    frame #10: 0x0000000180336de8 dyld`dyld3::MachOAnalyzer::forEachInitializer(Diagnostics&, dyld3::MachOAnalyzer::VMAddrConverter const&, void (unsigned int) block_pointer, void const*) const + 432
    frame #11: 0x00000001802edb38 dyld`dyld4::Loader::findAndRunAllInitializers(dyld4::RuntimeState&) const + 524
    frame #12: 0x00000001802f3f70 dyld`dyld4::JustInTimeLoader::runInitializers(dyld4::RuntimeState&) const + 36
    frame #13: 0x00000001802edf24 dyld`dyld4::Loader::runInitializersBottomUp(dyld4::RuntimeState&, dyld3::Array<dyld4::Loader const*>&) const + 220
    frame #14: 0x00000001802edf00 dyld`dyld4::Loader::runInitializersBottomUp(dyld4::RuntimeState&, dyld3::Array<dyld4::Loader const*>&) const + 184
    frame #15: 0x00000001802edf00 dyld`dyld4::Loader::runInitializersBottomUp(dyld4::RuntimeState&, dyld3::Array<dyld4::Loader const*>&) const + 184
    frame #16: 0x00000001802f1ab0 dyld`dyld4::Loader::runInitializersBottomUpPlusUpwardLinks(dyld4::RuntimeState&) const::$_1::operator()() const + 112
    frame #17: 0x00000001802ee0f0 dyld`dyld4::Loader::runInitializersBottomUpPlusUpwardLinks(dyld4::RuntimeState&) const + 380
    frame #18: 0x00000001803134dc dyld`dyld4::APIs::runAllInitializersForMain() + 464
    frame #19: 0x00000001802d5fa0 dyld`dyld4::prepare(dyld4::APIs&, dyld3::MachOAnalyzer const*) + 3192
    frame #20: 0x00000001802d4edc dyld`start + 1844
```
I think it has to do with the version of the C++ standard library that is getting linked. 
Adding `libcxx` and `libcxxabi` to the `nativeBuildInputs` and updating the `build_odin.sh` file to explicitly link `libcxxabi` allows it to run without segfaulting. (But thats just a workaround I'm using). 

EDIT: During more testing with the built binary, it also doesn't use the correct linker, which can cause problems if you are linking with Objective-C code/frameworks (not sure if this is a problem with how Odin is packaged with nix, or if its a problem that the Odin devs need to solve).

### Notify maintainers
@luc65r @Astavie @znaniye 

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 23.1.0, macOS 14.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by A1029384756**
## Description of changes
Moving to LLVM 18, the latest supported version of LLVM for the language. Fixes issues with Spall profiling when compiling without flags.

https://releases.llvm.org/18.1.0/docs/ReleaseNotes.html

Fixes #275919

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by matteo-pacini** [comment]
Result of `nixpkgs-review pr 319989` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>odin</li>
    <li>ols</li>
  </ul>
</details>

File: pkgs/development/compilers/odin/default.nix [hunk 1]
```diff
@@ -1,6 +1,6 @@
 { lib
 , fetchFromGitHub
-, llvmPackages_13
+, llvmPackages_18
```
Comment (superherointj) [inline pkgs/development/compilers/odin/default.nix:3]:
```suggestion
, llvmPackages_18
```
This is not the right way of overriding the version of a package.
This should be: `llvmPackages` or else the interface is going to change for overrides.

File: pkgs/development/compilers/odin/default.nix [hunk 2]
```diff
@@ -9,7 +9,7 @@
 }:
 
 let
-  llvmPackages = llvmPackages_13;
+  llvmPackages = llvmPackages_18;
```
Comment (superherointj) [inline pkgs/development/compilers/odin/default.nix:12]:
```suggestion
```

File: pkgs/development/compilers/odin/default.nix (hunk 1 above)
Comment (Astavie) [reply pkgs/development/compilers/odin/default.nix:3]: llvmPackages is still at version 17 though, while this PR wants to change it to version 18. do you mean we should specify the version inside `top-level/all-packages.nix`?

Comment (superherointj) [reply pkgs/development/compilers/odin/default.nix:3]: Or you could use `?`.

Comment (superherointj) [reply pkgs/development/compilers/odin/default.nix:3]: Yes, specify the version in `callPackage`.

Comment (superherointj) [reply pkgs/development/compilers/odin/default.nix:3]: Example from another package:
> pkgs/top-level/all-packages.nix
image

**Comment by superherointj** [comment]
Squash commits. Also format of commit message is wrong.

**Review by superherointj — APPROVED** [review]

**Pull request merged**